runtime/compare.c

```diff
@@ -99,6 +100,31 @@ static intnat compare_val(value v1, value v2, int total)
   return res;
 }
 
+static void poll_pending_actions(struct compare_stack* stk,
+                                 struct compare_item* next_stack_slot)
+{
+  if (caml_check_pending_actions()) {
+    value exn;
+    if (next_stack_slot == stk->stack) {
+      exn = caml_do_pending_actions_exn();
+    } else {
```

Comment:
Zero-sized blocks of local roots work fine as far as I can see.  (Function ` caml_do_local_roots` in runtime/roots.c)  So you don't need this smelly special case, the general case below should handle it just fine.

Reply:
Thanks, fixed.